### ISSUE TITLE
[What-If Tool] TF Records loading met with RequestNetworkError

### ISSUE THREAD

**Issue by zmjjmz** [tensorflow/tensorboard#3421]
## Environment information (required)

Results of diagnose_me script: see this [gist](https://gist.github.com/zmjjmz/00a58ac6acdf469bb05dffd2515816a2)

### Diagnostics

<details>
<summary>Diagnostics output</summary>

``````
--- check: autoidentify
INFO: diagnose_tensorboard.py version d515ab103e2b1cfcea2b096187741a0eeb8822ef

--- check: general
INFO: sys.version_info: sys.version_info(major=3, minor=5, micro=3, releaselevel='final', serial=0)
INFO: os.name: posix
INFO: os.uname(): posix.uname_result(sysname='Linux', nodename='wa1okdba002', release='4.9.0-11-amd64', version='#1 SMP Debian 4.9.189-3+deb9u2 (2019-11-11)', machine='x86_64')
INFO: sys.getwindowsversion(): N/A

--- check: package_management
INFO: has conda-meta: False
INFO: $VIRTUAL_ENV: '/home/u1/zach/proj/dataplayground3'
WARNING: Could not generate requirement for distribution -rpcio 1.22.0 (/home/u1/zach/proj/dataplayground3/lib/python3.5/site-packages): Parse error at "'-rpcio=='": Expected W:(abcd...)
WARNING: Could not generate requirement for distribution -ensorflow 1.14.0 (/home/u1/zach/proj/dataplayground3/lib/python3.5/site-packages): Parse error at "'-ensorfl'": Expected W:(abcd...)
WARNING: Could not generate requirement for distribution -cipy 1.3.1 (/home/u1/zach/proj/dataplayground3/lib/python3.5/site-packages): Parse error at "'-cipy==1'": Expected W:(abcd...)

--- check: installed_packages
INFO: installed: tensorboard==2.1.1
INFO: installed: tensorflow==2.1.0
INFO: installed: tensorflow-estimator==2.1.0
2020-03-23 19:19:26.258379: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'libnvinfer.so.6'; dlerror: libnvinfer.so.6: cannot open shared object file: No such file or directory
2020-03-23 19:19:26.258483: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'libnvinfer_plugin.so.6'; dlerror: libnvinfer_plugin.so.6: cannot open shared object file: No such file or directory
2020-03-23 19:19:26.258493: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:30] Cannot dlopen some TensorRT libraries. If you would like to use Nvidia GPU with TensorRT, please make sure the missing libraries mentioned above are installed properly.

--- check: tensorboard_python_version
Traceback (most recent call last):
  File "diagnose_tensorboard.py", line 470, in main
    suggestions.extend(check())
  File "diagnose_tensorboard.py", line 78, in wrapper
    result = fn()
  File "diagnose_tensorboard.py", line 246, in tensorboard_python_version
    from tensorboard import version
ImportError: cannot import name 'version'

--- check: tensorflow_python_version
WARNING: Limited tf.compat.v2.summary API due to missing TensorBoard installation.
WARNING: Limited tf.compat.v2.summary API due to missing TensorBoard installation.
WARNING: Limited tf.compat.v2.summary API due to missing TensorBoard installation.
WARNING: Limited tf.compat.v2.summary API due to missing TensorBoard installation.
WARNING: Limited tf.compat.v2.summary API due to missing TensorBoard installation.
WARNING: Limited tf.compat.v2.summary API due to missing TensorBoard installation.
WARNING: Limited tf.summary API due to missing TensorBoard installation.
INFO: tensorflow.__version__: '2.1.0'
INFO: tensorflow.__git_version__: 'v2.1.0-rc2-17-ge5bf8de'

--- check: tensorboard_binary_path
INFO: which tensorboard: b'/home/u1/zach/.local/bin/tensorboard\n'

--- check: addrinfos
socket.has_ipv6 = True
socket.AF_UNSPEC = <AddressFamily.AF_UNSPEC: 0>
socket.SOCK_STREAM = <SocketKind.SOCK_STREAM: 1>
socket.AI_ADDRCONFIG = 32
socket.AI_PASSIVE = 1
Loopback flags: 32
Loopback infos: [(<AddressFamily.AF_INET6: 10>, <SocketKind.SOCK_STREAM: 1>, 6, '', ('::1', 0, 0, 0)), (<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_STREAM: 1>, 6, '', ('127.0.0.1', 0))]
Wildcard flags: 1
Wildcard infos: [(<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_STREAM: 1>, 6, '', ('0.0.0.0', 0)), (<AddressFamily.AF_INET6: 10>, <SocketKind.SOCK_STREAM: 1>, 6, '', ('::', 0, 0, 0))]

--- check: readable_fqdn
INFO: socket.getfqdn(): '<REDACTED>'

--- check: stat_tensorboardinfo
INFO: directory: /tmp/.tensorboard-info
INFO: os.stat(...): os.stat_result(st_mode=16895, st_ino=2604, st_dev=65024, st_nlink=2, st_uid=30183, st_gid=30183, st_size=4096, st_atime=1585003314, st_mtime=1585004935, st_ctime=1585004935)
INFO: mode: 0o40777

--- check: source_trees_without_genfiles
INFO: tensorboard_roots (0): []; bad_roots (0): []
WARNING: Could not generate requirement for distribution -rpcio 1.22.0 (/home/u1/zach/proj/dataplayground3/lib/python3.5/site-packages): Parse error at "'-rpcio=='": Expected W:(abcd...)
WARNING: Could not generate requirement for distribution -ensorflow 1.14.0 (/home/u1/zach/proj/dataplayground3/lib/python3.5/site-packages): Parse error at "'-ensorfl'": Expected W:(abcd...)
WARNING: Could not generate requirement for distribution -cipy 1.3.1 (/home/u1/zach/proj/dataplayground3/lib/python3.5/site-packages): Parse error at "'-cipy==1'": Expected W:(abcd...)

--- check: full_pip_freeze
INFO: pip freeze --all:
absl-py==0.7.1
aenum==2.2.1
alembic==1.0.11
altgraph==0.16.1
anchor-exp==0.0.0.5
annoy==1.16.0
apipkg==1.5
appdirs==1.4.3
appnope==0.1.0
argh==0.26.2
asciimatics==1.11.0
astor==0.8.0
atomicwrites==1.3.0
attrs==19.1.0
avro==1.9.0
backcall==0.1.0
backports-abc==0.5
backports.csv==1.0.7
backports.shutil-get-terminal-size==1.0.0
backports.ssl-match-hostname==3.7.0.1
backports.weakref==1.0.post1
bcrypt==3.1.7
beautifulsoup4==4.8.0
bintrees==2.0.7
bleach==1.5.0
blis==0.2.4
bokeh==1.3.4
boto==2.49.0
boto3==1.10.23
botocore==1.13.23
bz2file==0.98
cached-property==1.5.1
cachetools==3.1.1
certifi==2019.6.16
cffi==1.13.2
chardet==3.0.4
ciso8601==2.1.1
clarifai==2.6.2
Click==7.0
click-plugins==1.1.1
cligj==0.5.0
cloudpickle==1.2.1
colorama==0.4.1
configobj==5.0.6
configparser==3.8.1
contextlib2==0.5.5
coverage==4.5.4
croniter==0.3.30
cryptography==2.8
cycler==0.10.0
cymem==2.0.2
Cython==0.29.13
cytoolz==0.10.0
dask==2.2.0
datasketch==1.4.7
ddt==1.2.1
decorator==4.4.0
defusedxml==0.6.0
descartes==1.1.0
dill==0.3.0
Distance==0.1.3
distributed==2.2.0
distro==1.4.0
docker==3.7.3
docker-compose==1.24.1
docker-pycreds==0.4.0
dockerpty==0.4.1
docopt==0.6.2
docutils==0.15.2
eli5==0.9.0
emoji==0.5.3
entrypoints==0.3
enum34==1.1.6
execnet==1.7.0
fastavro==0.22.3
fastcache==1.1.0
fasteners==0.15
fastparquet==0.3.2
feather-format==0.4.0
filechunkio==1.8
filelock==3.0.12
Fiona==1.8.6
flake8==3.7.8
Flask==1.1.1
Flask-Admin==1.5.3
Flask-Cache==0.13.1
Flask-Login==0.4.1
flask-swagger==0.2.14
Flask-WTF==0.14.2
fsspec==0.6.2
funcsigs==1.0.2
funcy==1.13
future==0.17.1
futures==3.1.1
gapic-google-cloud-datastore-v1==0.90.4
gapic-google-cloud-error-reporting-v1beta1==0.15.3
gapic-google-cloud-logging-v2==0.91.3
gapic-google-cloud-pubsub-v1==0.15.4
gapic-google-cloud-spanner-admin-database-v1==0.15.3
gapic-google-cloud-spanner-admin-instance-v1==0.15.3
gapic-google-cloud-spanner-v1==0.15.3
gapic-google-cloud-speech-v1beta1==0.15.3
gapic-google-cloud-vision-v1==0.90.3
gast==0.2.2
gensim==3.8.0
geographiclib==1.49
geopandas==0.5.1
geopy==1.20.0
gevent==1.4.0
gitdb==0.6.4
gitdb2==2.0.5
GitPython==3.0.0
google-api-core==1.14.2
google-api-python-client==1.8.0
google-auth==1.6.3
google-auth-httplib2==0.0.3
google-auth-oauthlib==0.4.1
google-cloud==0.34.0
google-cloud-bigquery==1.18.0
google-cloud-bigtable==0.34.0
google-cloud-core==1.0.3
google-cloud-datastore==1.9.0
google-cloud-dns==0.30.2
google-cloud-error-reporting==0.32.0
google-cloud-language==1.3.0
google-cloud-logging==1.12.1
google-cloud-monitoring==0.32.0
google-cloud-pubsub==0.45.0
google-cloud-resource-manager==0.29.2
google-cloud-runtimeconfig==0.29.2
google-cloud-spanner==1.10.0
google-cloud-speech==1.2.0
google-cloud-storage==1.18.0
google-cloud-translate==1.6.0
google-cloud-vision==0.38.1
google-gax==0.16.0
google-pasta==0.1.7
google-resumable-media==0.3.2
googleapis-common-protos==1.6.0
gql==0.2.0
grandalf==0.6
graphql-core==1.1
graphviz==0.11.1
greenlet==0.4.15
grpc-google-iam-v1==0.12.3
grpcio==1.27.2
gunicorn==19.9.0
h5py==2.9.0
HeapDict==1.0.0
hpsklearn==0.1.0
html-validator==1.0.1
html5lib==0.9999999
httplib2==0.13.1
humanize==0.5.1
hyperopt==0.2.3
idna==2.7
ImageHash==4.0
imageio==2.5.0
implicit==0.3.8
importlib-metadata==0.19
inflect==2.1.0
ipaddress==1.0.22
ipdb==0.13.2
ipykernel==5.1.2
ipython==7.7.0
ipython-genutils==0.2.0
ipywidgets==7.5.1
itsdangerous==1.1.0
jax==0.1.55
jaxlib==0.1.37
jedi==0.15.0
Jinja2==2.10.1
jmespath==0.9.4
joblib==0.13.2
json2parquet==0.0.28
jsonpath-ng==1.4.3
jsonschema==2.6.0
jupyter==1.0.0
jupyter-client==5.3.1
jupyter-console==6.0.0
jupyter-contrib-core==0.3.3
jupyter-contrib-nbextensions==0.5.1
jupyter-core==4.5.0
jupyter-highlight-selected-word==0.2.0
jupyter-latex-envs==1.4.6
jupyter-nbextensions-configurator==0.4.1
jupyternotify==0.1.15
jupyterthemes==0.20.0
Keras==2.2.4
Keras-Applications==1.0.8
Keras-Preprocessing==1.1.0
kerassurgeon==0.1.3
kiwisolver==1.1.0
langid==1.1.6
lesscpy==0.13.0
lightfm==1.15
lime==0.1.1.36
llist==0.6
llvmlite==0.29.0
locket==0.2.0
lockfile==0.12.2
lru-dict==1.1.6
lxml==4.4.1
macholib==1.11
Mako==1.1.0
Markdown==3.1.1
MarkupSafe==1.1.1
matplotlib==3.0.3
mccabe==0.6.1
memory-profiler==0.55.0
mistune==0.8.4
mock==3.0.5
modulegraph==0.17
monotonic==1.5
more-itertools==7.2.0
msgpack==0.6.1
msgpack-numpy==0.4.4.3
msgpack-python==0.5.6
munch==2.3.2
murmurhash==1.0.2
mysql==0.0.2
mysqlclient==1.4.5
nanotime==0.5.2
nbconvert==5.6.0
nbformat==4.4.0
networkx==2.2
nltk==3.4.4
nose==1.3.7
notebook==6.0.0
np-utils==0.5.10.0
numba==0.45.1
numexpr==2.6.9
numpy==1.17.3
nvidia-ml-py3==7.352.0
oauth2client==4.1.3
oauthlib==3.1.0
olefile==0.46
opencv-python==4.1.0.25
opt-einsum==3.1.0
ordereddict==1.1
packaging==19.1
pandarallel==1.4.5
pandas==0.25.0
pandoc==1.0.2
pandocfilters==1.4.2
paramiko==2.6.0
parquet==1.2
parso==0.5.1
partd==1.0.0
pathlib==1.0.1
pathlib2==2.3.4
pathspec==0.5.9
pathtools==0.1.2
patsy==0.5.1
pbr==5.4.2
pexpect==4.7.0
phash==0.2.0
pickleshare==0.7.5
Pillow==6.1.0
pip==19.3.1
plac==1.1.0
plotly==4.5.0
pluggy==0.12.0
ply==3.11
pprint==0.1
preshed==2.0.1
prometheus-client==0.7.1
promise==2.3
prompt-toolkit==2.0.9
proto-google-cloud-datastore-v1==0.90.4
proto-google-cloud-error-reporting-v1beta1==0.15.3
proto-google-cloud-logging-v2==0.91.3
proto-google-cloud-pubsub-v1==0.15.4
proto-google-cloud-spanner-admin-database-v1==0.15.3
proto-google-cloud-spanner-admin-instance-v1==0.15.3
proto-google-cloud-spanner-v1==0.15.3
proto-google-cloud-speech-v1beta1==0.15.3
proto-google-cloud-vision-v1==0.90.3
protobuf==3.9.1
protobuf-to-dict==0.1.0
protobuf3-to-dict==0.1.5
psutil==5.6.3
psycopg2==2.8.3
ptyprocess==0.6.0
py==1.8.0
py2app==0.19
pyarrow==0.14.1
pyasn1==0.4.6
pyasn1-modules==0.2.6
pycodestyle==2.5.0
pycparser==2.19
pyemd==0.5.1
pyenchant==2.0.0
pyfiglet==0.8.post1
pyflakes==2.1.1
Pygments==2.4.2
PyInstaller==3.5
pyLDAvis==2.1.2
pylru==1.2.0
pymongo==3.8.0
Pympler==0.7
PyMySQL==0.9.3
PyNaCl==1.3.0
pyparsing==2.4.2
pyproj==2.2.1
pyrsistent==0.15.4
pytest==5.0.1
pytest-cov==2.7.1
pytest-forked==1.0.2
pytest-xdist==1.29.0
python-daemon==2.2.3
python-dateutil==2.8.0
python-editor==1.0.4
python-nvd3==0.15.0
python-slugify==3.0.3
python-snappy==0.5.4
pytz==2019.2
PyWavelets==1.0.3
PyYAML==5.3
pyzmq==18.1.0
qtconsole==4.5.2
redis==3.3.6
regex==2019.6.8
requests==2.23.0
requests-oauthlib==1.3.0
retrying==1.3.3
rsa==4.0
ruamel.yaml==0.15.100
Rx==1.6.1
s3transfer==0.2.1
sagemaker==1.44.0
sagemaker-container-support==1.1.3
scandir==1.10.0
schema==0.7.0
scikit-image==0.15.0
scikit-learn==0.21.3
scikit-surprise==1.1.0
scipy==1.4.1
seaborn==0.9.0
Send2Trash==1.5.0
sentry-sdk==0.13.5
setproctitle==1.1.10
setuptools==41.0.1
Shapely==1.6.4.post2
shortuuid==0.5.0
Sift==5.0.1
simplegeneric==0.8.1
singledispatch==3.4.0.3
six==1.12.0
smart-open==1.8.4
smmap==0.9.0
smmap2==2.0.5
sortedcontainers==2.1.0
soupsieve==1.9.2
spacy==2.1.8
sql==0.4.0
SQLAlchemy==1.3.6
srsly==0.0.7
statsmodels==0.10.1
statutils==0.0.1
subprocess32==3.5.4
tabulate==0.8.3
tblib==1.4.0
tensorboard==2.1.1
tensorflow==2.1.0
tensorflow-estimator==2.1.0
tensorflow-hub==0.5.0
tensorflow-serving-api==1.14.0
termcolor==1.1.0
terminado==0.8.2
testpath==0.4.2
text-unidecode==1.2
texttable==0.9.1
thinc==7.0.8
thrift==0.11.0
thriftpy==0.3.9
timeout-decorator==0.4.1
toml==0.10.0
toolz==0.10.0
torch==1.3.1
tornado==6.0.3
tox==3.13.2
tqdm==4.33.0
traitlets==4.3.2
treelib==1.5.5
typing==3.7.4
ujson==1.35
unicodecsv==0.14.1
Unidecode==1.1.1
uritemplate==3.0.1
urllib3==1.24.3
virtualenv==16.7.2
wandb==0.8.28
wasabi==0.2.2
watchdog==0.9.0
wcwidth==0.1.7
webencodings==0.5.1
websocket-client==0.56.0
weighted-levenshtein==0.2.1
Werkzeug==0.15.5
wheel==0.33.4
widgetsnbextension==3.5.1
witwidget==1.6.0
wrapt==1.11.2
WTForms==2.2.1
xmltodict==0.12.0
yapf==0.28.0
zc.lockfile==2.0
zict==1.0.0
zipp==0.5.2
zope.deprecation==4.4.0

``````

</details>

## Issue description

The tensorboard server is running as follows:

`tensorboard --logdir fts_logdir/ --alsologtostderr`

Upon pointing the What-If tool to the TF record file I'd created, I get the following error:

```
Request for datapoint load failed: RequestNetworkError: RequestNetworkError: 400 at /data/plugin/whatif/examples_from_path?examples_path=%2Fhome%2Fu1%2F<path_redacted>%2Ffts_logging%2Fwit_test.tfrecord&max_examples=1000&sampling_odds=1&sequence_examples=false
```

To create the TF records (_if_ that's the issue), I have a DataFrame with some features and put it through the following function (plagiarized from other What-If tool notebooks):

```
import tensorflow as tf
from tqdm import tqdm_notebook

def df_to_examples(df, columns=None):
    examples = []
    if columns == None:
        columns = df.columns.values.tolist()
    for index, row in tqdm_notebook(df.iterrows(), total=len(df)):
        example = tf.train.Example()
        for col in columns:
            if np.issubdtype(df[col].dtype, np.dtype(int)) or np.issubdtype(df[col].dtype, np.dtype(bool)):
                example.features.feature[col].int64_list.value.append(int(row[col]))
            elif np.issubdtype(df[col].dtype, np.dtype(float)):
                example.features.feature[col].float_list.value.append(row[col])
            elif row[col] == row[col]:
                print("Col {0} has dtype {1}, treating as bytes_list".format(
                    col, df[col].dtype
                )) # NOTE: in this case, I've never hit this issue
                example.features.feature[col].bytes_list.value.append(row[col])
        examples.append(example)
    return examples

examples_list = df_to_examples(df, columns=features)
tfrecord_fn = 'wit_test.tfrecord'
writer = tf.io.TFRecordWriter(tfrecord_fn)
for ex in tqdm_notebook(examples_list):
    writer.write(ex.SerializeToString())
```

The only other reference to this seems to be from when a CSV loader wasn't ready yet, so I'm assuming it's plausible that I've somehow screwed up creating these TF records, but it's not clear to me how.

**Comment by stephanwlee** [comment]
Can you please attach the result of diagnomse_me script?

Instructions:
> Please run `diagnose_tensorboard.py` (link below) in the same
environment from which you normally run TensorFlow/TensorBoard, and
paste the output here:
> https://raw.githubusercontent.com/tensorflow/tensorboard/master/tensorboard/tools/diagnose_tensorboard.py

**Comment by zmjjmz** [comment]
It's in the gist I linked, updated my post to make that more clear

**Comment by stephanwlee** [comment]
@jameswex could you take a look at this? It looks like 2.1.1 is the last release without the dynamic plugin but, regardless, the problem should be reproducible in the new plugin, too.

**Comment by jameswex** [comment]
Thanks @zmjjmz . Do you see any python errors/stack traces in the Tensorboard's log output to stderr?
Also are you able to provide your tfrecord file, if it doesn't contain sensitive data?

I just quickly tested Tensorboard 2.1.1 WIT with my own test tfrecord file and saw no issues. So we'll have to see if its related to your file, or your setup perhaps.

**Comment by zmjjmz** [comment]
Hey again @jameswex, I cannot share the data itself. The only thing I see from the tensorboard's output is an (I think) unrelated warning

```
W0323 19:09:08.049231 139812368647936 deprecation.py:323] From /home/u1/zach/.local/lib/python2.7/site-packages/tensorboar
d/plugins/interactive_inference/utils/platform_utils.py:143: tf_record_iterator (from tensorflow.python.lib.io.tf_record)
is deprecated and will be removed in a future version.
Instructions for updating:
Use eager execution and:
`tf.data.TFRecordDataset(path)`
```

**Comment by zmjjmz** [comment]
FWIW: This is the same model I was trying to get working with the WITWidget (https://github.com/PAIR-code/what-if-tool/issues/61). I imagine these issues are related?

**Comment by jameswex** [comment]
Not sure if they're related as the WIT issue was at inference-time, and this issue is at example-loading time, which are separate operations.

Are you able to load up your tf.Examples from your tfrecords on disk using logic similar to https://github.com/PAIR-code/what-if-tool/blob/master/utils/platform_utils.py#L65?

I'm wondering if there is something wrong with the file as it is persisted on disk.

**Comment by zmjjmz** [comment]
```
tfrecord_fn = 'fts_logdir/wit_test.tfrecord'
writer = tf.io.TFRecordWriter(tfrecord_fn)
for ex in pred_df_ex:
    writer.write(ex.SerializeToString())

record_iterator = tf.compat.v1.python_io.tf_record_iterator(
    path='fts_logdir/wit_test.tfrecord')
ex = list(record_iterator)

parsed_ex = [tf.train.Example.FromString(example) for example in ex]
assert(all([
    (parsed_ex[i].features.feature['<feature_name>'].int64_list.value[0] ==
     pred_df_ex[i].features.feature['<feature_name>'].int64_list.value[0]) for i in range(len(parsed_ex))]))
```

I'm fairly certain this test replicates the core parts of that example here, and it works fine.

**Comment by jameswex** [comment]
Thanks for the quick responses. I just created a tfrecord file for the uci census data from https://colab.sandbox.google.com/github/pair-code/what-if-tool/blob/master/WIT_Model_Comparison.ipynb, tried it in my TB 2.1.1 WIT to verify it could load, and attached it here. To narrow down the problem, is your TensorBoard WIT able to load that data file without error?

[wit_test_uci.tfrecord.zip](https://github.com/tensorflow/tensorboard/files/4378206/wit_test_uci.tfrecord.zip)

**Comment by zmjjmz** [comment]
Ok, sorry about that -- it ended up being a weird permissions issue on my end. That said, I'm now encountering the same issue [I was encountering before](https://github.com/PAIR-code/what-if-tool/issues/61#issuecomment-602932032) with the inability to provide the tensor name properly for a dictionary input, although I don't think I can insert a custom predict function in this case. Let me know if I should make a separate ticket for that, or if there's a workaround in the meantime.

**Comment by jameswex** [comment]
With WIT in TensorBoard, it currently only works for models served by TFServing that use the ClassificationRequest, RegressionRequest, or PredictionRequest APIs through the TF serving PredictionService, where the input is provided as a tensor containing a serialized Example proto.

For your case where you provide the input as a dictionary, there is no way to use the TensorBoard version of WIT. But as you saw in https://github.com/PAIR-code/what-if-tool/issues/61, WIT still works for this case in notebook mode, through use of a custom predict function.

Feel free to open a feature request on WIT at https://github.com/PAIR-code/what-if-tool/issues/new to track your feature request for being able to use WIT in TensorBoard with your served model setup.
